### ISSUE TITLE
Interleaved parallel job output in LocalEnvironment

### ISSUE THREAD

**Issue by schuhschuh** [openmole/openmole#52]
When the LocalEnvironment is used, the output of jobs that are run in parallel is interleaved in the console output. This is different from when the jobs are submitted to a remote execution node (for understandable reasons). It would be nice if also in case of the LocalEnvironment, the output of each job would be printed separately in one junk.

**Comment by romainreuillon** [comment]
That would be great but it is non trivial. It will require to override the default printstream of System.out and System.err and check if the calling thread is a local execution environment thread. If so it would store the message in a stringbuilder (stored in a weakhashmap thread->stringbuilder to avoid memory leak) if not it would display it. At the end of a job execution, the thread would look into the weakhashmap and display the output.

**Comment by schuhschuh** [comment]
That sounds complicated indeed. Would it not be easier to only redirect the output of the local execution environment threads to separate streams at the time they are spawned? I suppose it's one of your code that spawns these threads.

**Comment by romainreuillon** [comment]
It is, but in the JVM there is no per-thread stdout. System.out is a global variable for the all JVM.

**Comment by schuhschuh** [comment]
I see. One simpler alternative might be a new environment instead for longer running local tasks (not just a simple ScalaTask or EmptyTask) which creates a separate process? Because for these you can get the stdout and stderr streams easily. And it would mimic the task execution on a remote machine.

But I suppose this can already be done using a SSHEnvironment with "localhost"... ?

**Comment by romainreuillon** [comment]
After a bit of thinking I think the lightweight solution is feasible and it would have the advantage be able to redirect the output on a mole execution bases which would be good for the graphical user interface. I'll give it a try tomorrow.

**Comment by schuhschuh** [comment]
:) As long as it serves a greater purpose than just to avoid the interleaved output, I'm all for it.

**Comment by romainreuillon** [comment]
It was no picnic but, I just pushed the feature. You test is using:
```val env = LocalEnvironment(interleave = true)```

**Comment by schuhschuh** [comment]
Nice! I think you meant ```deinterleave = true```, though...

**Comment by romainreuillon** [comment]
Yes, that it.

**Comment by schuhschuh** [comment]
I think you also want to set this optional argument to "false" by default such that existing OpenMOLE scripts compile and run as before.

**Comment by schuhschuh** [comment]
Has this been tested? Because actually it doesn't seem to work. I am not sure yet if that's because of my WF or because of how it's implemented. Will it cause problems if my tasks have different instances of a LocalEnvironment ? If that shouldn't cause a problem, then I must say this feature isn't working yet.

**Comment by romainreuillon** [comment]
It has been tested. Could you produce a simple workflow that fails so I can fix it?

**Comment by schuhschuh** [comment]
I hope so, I'll get back to this once I resolved a number other problems I have.

**Comment by schuhschuh** [comment]
I tried with the following script. When I set ```nbThreads = 1```, the output of the main tasks is not interleaved in either case, of course. But with a higher number of threads/parallel main tasks, the output remains interleaved.

But what I actually wanted to see is the output of the ```post``` tasks. For some reasons these are not being executed... I've had similar issue before. What am I missing in the WF definition ?

```scala
import org.openmole.core.workflow.data.Prototype

val env = LocalEnvironment(1, deinterleave = true)

val i = Val[Int]

def exploration = ExplorationTask(i in (1 to 5))

def main =
  Capsule(
    ScalaTask(
      s"""
        | for (line <- 1 to 9)
        |   println(f"main $$line%02d: i = $$i")
      """.stripMargin
    ) set (
      name    := "main",
      inputs  += i,
      outputs += i
    )
  )

def post(id: String, i: Prototype[Int]) =
  ScalaTask(
    s"""
      | for (line <- 1 to 9)
      |   println(f"$id $$line%02d: i = $$i")
    """.stripMargin
  ) set (
    name   := id,
    inputs += i
  )

val ex = ((exploration -< (main on env)) + (main -- post("post1", i)) + (main -- post("post2", i)) + (main -- post("post3", i)) + (main -- post("post4", i))).start
ex.waitUntilEnded
```

Also, with ```nbThreads = 1```, the output of the first executed task is not enclosed by the separating lines:
```
ex: org.openmole.core.workflow.mole.MoleExecution = org.openmole.core.workflow.mole.MoleExecution@7172525e
main 01: i = 1
main 02: i = 1
main 03: i = 1
main 04: i = 1
main 05: i = 1
main 06: i = 1
main 07: i = 1
main 08: i = 1
main 09: i = 1
----------------------------Output of local execution----------------------------
main 01: i = 5
main 02: i = 5
main 03: i = 5
main 04: i = 5
main 05: i = 5
main 06: i = 5
main 07: i = 5
main 08: i = 5
main 09: i = 5
---------------------------------------------------------------------------------
...
```

**Comment by romainreuillon** [comment]
That's because the default local environment is not desinterleaved, don't you think?

**Comment by schuhschuh** [comment]
But I explicitly instantiate an ```env``` for the ```main``` task with ```deinterleave = true```. This is for some reason the only task that is being executed in this workflow.

**Comment by schuhschuh** [comment]
It's in particular strange that the output is different depending on the number of threads (```nbThreads```).

**Comment by romainreuillon** [comment]
Yes you're right, I didn't see it was a capsule.

**Comment by romainreuillon** [comment]
This should be fixed in: 4c38fab5908ea9bb6278d1505b30f5eee3aa0042.

NB: in the example, the main capsule should be assigned to a value (val instead of def) otherwise the post task is not connected to the workflow.

**Comment by schuhschuh** [comment]
Ah, yes, it should be a val... :S This works now just as intended.
